MetaMask/metamask-mobile
# feat: snaps controllers integration (Flask Only)

## **Description**

This pr enables snaps to be installed and executed on mobile. This change will be code fenced for flask only. We achieve this by doing the following..
- integrating the [@metamask/snaps-controllers](https://github.com/MetaMask/snaps/tree/main/packages/snaps-controllers)
- integrating [@metamask/snaps-rpc-methods](https://github.com/MetaMask/snaps/tree/main/packages/snaps-rpc-methods)
- adding handlers for the snaps rpc methods into the [BackgroundBridge](https://github.com/MetaMask/metamask-mobile/pull/7942/files#diff-e39cb6f200b31535fa6c6cebdc19d7721a389d27c1749ce705d527c84117349cR324).js file
- implementing the [InstallSnapApproval](https://github.com/MetaMask/metamask-mobile/pull/7942/files#diff-30d7cdcba3396fa012d38e24590073939e59a76a12e01fe30de7fa662c9c5cfdR15) flow
- create a custom [detectSnapLocation function](https://github.com/MetaMask/metamask-mobile/pull/7942/files#diff-5fdaee36cd7b672d53e8df96b9c0118531bf47b75d4154777b53f2522ce34c96R20)
- implementing a custom [fetch function](https://github.com/MetaMask/metamask-mobile/pull/7942/files#diff-492ac0406eca60eae1445537f5a2ff4f9a9b8a99f992c95276542ed8adde5a4eR39) that we pass to the snaps controller
- For npm hosted snaps, we will fetch them using the custom [NPMLocation](https://github.com/MetaMask/metamask-mobile/pull/7942/files#diff-82d4be138d97cff4dec8f1c78898939c31103d8aab98930cc9672099b9572e53R149) class. This classic fetches (using the custom fetch function mentioned above), downloads the compressed snap code and then decompresses and reads the snap code using the [RNTar native module that was implemented/merged in this PR](https://github.com/MetaMask/metamask-mobile/pull/7955).

All of these changes have been implemented and tested in [this PR](https://github.com/MetaMask/metamask-mobile/pull/7526). This PR updates that logic and brings it into main behind a code fence. 

## **Guide for reviewers**
- [ ] All modified code should be wrapped inside a code fence `///: BEGIN:ONLY_INCLUDE_IF(snaps)...///: END:ONLY_INCLUDE_IF`
- [ ] all new files should be wrapped in a code fence `///: BEGIN:ONLY_INCLUDE_IF(snaps)...///: END:ONLY_INCLUDE_IF`
- The most important files for to review are...
    1. Engine.ts
    2. BackgroundBridge.js
    3.  useApprovalRequest.ts
    4. EngineService.ts
    5. npm.ts
    6. location.ts
    7. fetch.ts

## **Related issues**

Progresses: https://github.com/MetaMask/accounts-planning/issues/143

## **Manual testing steps**

1. checkout this branch
8. run `yarn setup`
9. open the `.js.env`
10. modify the `METAMASK_BUILD_TYPE` to  `export METAMASK_BUILD_TYPE="flask"`
11. in your terminal run `source .js.env` and then run `yarn watch:clean`
12. in xcode select the flask variant  and press play
13. open http://metamask.github.io/snaps/test-snaps/1.1.0 in the browser

<img width="318" alt="Screenshot 2023-12-08 at 12 04 41 AM" src="https://github.com/MetaMask/metamask-mobile/assets/22918444/1a4a9cda-de34-4257-92a2-8705e5253720">

14. create a wallet
15. navigate to the browser
16. open this url: https://metamask.github.io/snaps/test-snaps/1.1.0/
17. install a snap, I usually install the bip32/bip44 snap
18. the install flow should appear with 3 steps including the a success screen
19. click the `Get Public key` button on the snap site, you should get a result
20. navigate to `settings/snaps`
21. you should see a list of all your installed snaps

### Checking code fencing
1. reset your .js.env file so that the METAMASK_BUILD_TYPE is set to `main`
2. restart your watchman server and refresh the app
3. the `snaps` section in the settings list should no longer exist
4. installing a snap should fail.

## **Screenshots/Recordings**



### **After**

https://github.com/MetaMask/metamask-mobile/assets/22918444/d0c957ec-e95e-4a4b-8526-d3d98586e671


## **Pre-merge author checklist**


## **Pre-merge reviewer checklist**
